omegatrade-app/backend/app/models/simulation.model.js

```diff
@@ -0,0 +1,83 @@
+'user strict';
+const database = require('./../config/database.js');
+const { v4: uuidv4 } = require('uuid');
+const Simulation = function () { };
+
+Simulation.getAll = async function (cb) {
+    try {
+        const [result] = await database.run({ sql: 'select sml.sId as sId,sml.companyId as companyId, sml.status as status, cy.companyName as companyName, cy.companyShortCode as companyShortCode from simulations sml LEFT JOIN companies cy ON sml.companyId = cy.companyId LIMIT 3', json: true });
+        cb(null, result)
+    } catch (error) {
+        cb(error, null)
+    }
+}
+
+Simulation.findById = async function (param, cb) {
+    try {
+        const sId = param.sId
+        const [ result ] = await database.run({
+            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where sId = @sId',
+            params: {
+                sId: sId
+            },
+            json: true
+        });
+        cb(null, result)
+        
+    } catch (error) {
+        cb(error, null)
+    }
+}
+
+Simulation.findByCompanyId = async function (companyId, sid) {
+    try {
+        const [result] = await database.run({
+            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where companyId = @companyId and sid = @sid',
+            params: {
+                companyId: companyId,
+                sid: sid
+            },
+            json:true
+        });
+        return result;
+    } catch (error) {
+        throw ("ERROR:", error);
```

Comment:
Here and elsewhere: The error handling seems to be a little inconsistent, with sometimes returning only a very generic error message, and sometimes including the entire underlying error. The preceding error message ( `ERROR:`, `Error:`) is also not always consistent, and in these cases it also does not provide anything useful above just letting the underlying error bubble up automatically. (Prepending the underlying error with `ERROR:` is not useful for debugging and also does not help the user understand what they did wrong)

So in this case (and many of the other functions) the entire try-catch block can be removed. Otherwise, the catch clause should do something useful with the error, such as adding a useful context message to the underlying error, or only returning a more generic contextual error message. This should preferably also be consistent across the application for the different layers. So for example (feel free to choose the approach that you think is best, but try to stick to a consistent scheme):
* Errors in the data layer always return the underlying error unchanged. In that case the current try-catch blocks can be removed from the data access layer.
* Errors in the data layer always return the underlying error with a context error message prepended. So in this specific function that would be something like `throw ("Failed to find company by id", error);`.
* Errors in the data layer always return a contextual error message without any information from the underlying error to prevent technical information from bubbling up. In this specific function that could be something like `throw Error("Error finding a company by id");`

Reply:
@olavloite 
I would like to remove the callback functions in models and the callers in controllers
For example  
in simulation controller

> try{
const simulation = await findById(simulationId);
return res.status(200).json({ success: true, data:simulation });
}catch (error){
     return res.status(500).json({ success: false, message: error });
}

and in simulation model

> simulation.findById = async function(simulationId){
	try	{const sId = param.sId
        const [ result ] = await database.run({
            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where sId = @sId',
            params: {
                sId: simulationId
            },
            json: true
        });
	return result;
        },catch(error){
        throw Error("Error finding a company by id")  // third suggestion
        }
};

In this approach, All Errors will be handled from a single try-catch block of the controller and any incoming error from the data layer can be given in the response. 
In case of the custom validation message (where I should not return technical information to frontend) will return messages like `user already exists` in case of the already registered user and genric message like `Authentication failed` in case of any failure in the authentication.
And for Debugging the error in order to know the technical information, I've planned to integrate logger service.
please suggest

Reply:
That sounds very reasonable to me. If I understand it correctly, it would mean that:
1. All the try-catch blocks in the data layer can be removed and any exception will be allowed to bubble up to the controller.
2. All controller functions will have a try-catch block that catch any data-layer related errors. You might need some additional logic in the controller in some very specific cases when you want to return a specific error for specific data problems. So for example for the case that you mentioned above, `User already exists`, you might want to add a check that the error code that was returned by the data layer actually is `AlreadyExists. If it is, you return the error message `User already exists`, and in other cases you could return a slightly more generic error message (e.g. `Something went wrong while creating a new user. Please check that the data you entered are valid.`).

Reply:
The above suggestion has been added and integrated with the logger service for debugging technical errors.